File: services/hub/cmd/brain/workerescalation_test.go
package main

import (
	"strings"
	"testing"
)

const validEscalationJSON = `{"type":"worker-escalation","status":"blocked","reason":"publishing requires write authority","requiredAuthorityOrDecision":"authorize a release-capable worker","changed":false,"nextAction":"review the artifact, then redispatch publishing"}`

func TestWorkerEscalationContractAndParser(t *testing.T) {
	if !strings.Contains(workerEscalationContract, "wks-escalation") ||
		!strings.Contains(workerEscalationContract, "requiredAuthorityOrDecision") ||
		!strings.Contains(workerEscalationContract, "wks-result") {
		t.Fatalf("contract does not advertise the fixed shape and result compatibility:\n%s", workerEscalationContract)
	}
	out := readWorkerEscalation("cannot publish\n\n```wks-escalation\n" + validEscalationJSON + "\n```")
	if out == nil || out.Error != "" || !strings.Contains(out.JSON, `"changed": false`) {
		t.Fatalf("valid escalation was not preserved: %+v", out)
	}
}

func TestWorkerEscalationMalformedAndAbsent(t *testing.T) {
	bad := readWorkerEscalation("```wks-escalation\n{\"type\":\"worker-escalation\"}\n```")
	if bad == nil || bad.Error == "" || bad.JSON != "" {
		t.Fatalf("malformed escalation was silently accepted: %+v", bad)
	}
	if got := readWorkerEscalation("I cannot publish because I only have read-only authority."); got != nil {
		t.Fatalf("ordinary prose refusal became an escalation: %+v", got)
	}
}

File: apps/desktop/src/main/shared/workerEscalation.test.ts
import { describe, expect, it } from 'vitest';
import {
  buildWorkerEscalationContract,
  readWorkerEscalation,
  WORKER_ESCALATION_FENCE,
} from './workerEscalation';

const valid = {
  type: 'worker-escalation',
  status: 'blocked',
  reason: 'Publishing requires authority I do not have.',
  requiredAuthorityOrDecision: 'Confirm whether to publish the release.',
  changed: false,
  nextAction: 'Review the local artifact, then dispatch a publisher with release authority.',
} as const;

describe('worker escalation terminal contract', () => {
  it('advertises the fixed shape and its relationship to optional wks-result', () => {
    const text = buildWorkerEscalationContract();
    expect(text).toContain(`\`\`\`${WORKER_ESCALATION_FENCE}`);
    expect(text).toContain('requiredAuthorityOrDecision');
    expect(text).toContain('"changed": false');
    expect(text).toContain('emit `wks-result` when you complete');
  });

  it('parses and normalizes a valid terminal escalation', () => {
    const out = readWorkerEscalation(
      `I cannot publish safely.\n\n\`\`\`${WORKER_ESCALATION_FENCE}\n${JSON.stringify(valid)}\n\`\`\``,
    );
    expect(out?.value).toEqual(valid);
    expect(out?.json).toContain('"requiredAuthorityOrDecision"');
    expect(out?.error).toBeUndefined();
  });

  it('rejects malformed JSON and wrong shapes instead of accepting them', () => {
    expect(readWorkerEscalation(`\`\`\`${WORKER_ESCALATION_FENCE}\n{nope}\n\`\`\``)?.error).toMatch(
      /not valid JSON/,
    );
    expect(
      readWorkerEscalation(
        `\`\`\`${WORKER_ESCALATION_FENCE}\n${JSON.stringify({ ...valid, changed: 'no' })}\n\`\`\``,
      )?.error,
    ).toBe('changed: expected boolean');
  });

  it('returns null for ordinary prose, including a plain refusal', () => {
    expect(
      readWorkerEscalation('I cannot publish because I only have read-only authority.'),
    ).toBeNull();
  });
});
